### PR TITLE
Move to experiment-based Hydra config in L-MGN example

### PR DESCRIPTION
**Pull request by Alexey-Kamenev**
# Modulus Pull Request

## Description

* moved to Hydra experiment config which simplifies configuring different dataset experiments in L-MGN.
* The example now supports 15+ datasets that are used in the original paper.
* Unified training and inference script configuration which are now both experiment-driven which guarantees coherency.
* Refactored logging since W&B is not available anymore.

Results on different datasets: 
* Water:
    
* Water 3D:
    
* Sand:
    
* Water ramps:
    

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/datapipes/gnn/lagrangian_dataset.py [hunk 1]
```diff
@@ -150,24 +155,30 @@ def __init__(
             self.acc_std = torch.tensor(metadata["acc_std"]).reshape(1, self.dim)
 
         # override from config
+        # TODO(akamenev): this is an unconditional overwrite of values that
```
Comment (ktangsali) [inline modulus/datapipes/gnn/lagrangian_dataset.py:158]: Is this intentional or a leftover?

**Review by ktangsali — APPROVED** [review]
Great organization! Thanks!

File: examples/cfd/lagrangian_mgn/README.md [hunk 2]
```diff
@@ -65,54 +71,88 @@ a small amount of noise is added during training.
...
-directory.
+## Inference
+
+The inference script, `inference.py`, also supports Hydra configuration, ensuring
+consistency between training and inference runs.
+
+Once the model is trained, run the following command:
+
+```bash
+python inference.py +experiment=water \
```
Comment (hakhondzadeh) [inline examples/cfd/lagrangian_mgn/README.md:145]: is there a way to specify sample id for the inference? So that we can try running the model on different data (including the ones used for training) and see the output?

File: examples/cfd/lagrangian_mgn/train.py [hunk 3]
```diff
@@ -14,98 +14,79 @@
...
         assert DistributedManager.is_initialized()
         self.dist = DistributedManager()
-        self.amp = cfg.amp
-        self.radius = cfg.radius
-        self.dt = cfg.dt
+
+        self.dt = cfg.data.train.dt
         self.dim = cfg.dim
-        self.gravity = torch.zeros(self.dim, device=self.dist.device)
-        self.gravity[-1] = -9.8
```
Comment (hakhondzadeh) [inline examples/cfd/lagrangian_mgn/train.py:51]: How is gravity considered now?

File: examples/cfd/lagrangian_mgn/train.py [hunk 4]
```diff
@@ -248,27 +203,30 @@ def main(cfg: DictConfig) -> None:
         mean_loss_pos = sum(loss_pos_list) / len(loss_pos_list)
         mean_loss_vel = sum(loss_vel_list) / len(loss_vel_list)
         mean_loss_acc = sum(loss_acc_list) / len(loss_acc_list)
```
Comment (hakhondzadeh) [inline examples/cfd/lagrangian_mgn/train.py:205]: Why not use np?

File: examples/cfd/lagrangian_mgn/train.py [hunk 5]
```diff
@@ -248,27 +203,30 @@ def main(cfg: DictConfig) -> None:
         mean_loss_pos = sum(loss_pos_list) / len(loss_pos_list)
```
Comment (hakhondzadeh) [inline examples/cfd/lagrangian_mgn/train.py:203]: mean_loss is only used for logging and not for training, right? Then, maybe rename it so it doesn't get confused with the loss being used for training?

File: examples/cfd/lagrangian_mgn/train.py [hunk 6]
```diff
@@ -202,38 +171,24 @@ def backward(self, loss):
...
-
-@hydra.main(version_base="1.3", config_path="conf", config_name="config_2d")
+@hydra.main(version_base="1.3", config_path="conf", config_name="config")
 def main(cfg: DictConfig) -> None:
     # initialize distributed manager
     DistributedManager.initialize()
     dist = DistributedManager()
 
+    init_python_logging(cfg, dist.rank)
+    logger.info(f"Config summary:\n{OmegaConf.to_yaml(cfg, sort_keys=True)}")
```
Comment (hakhondzadeh) [inline examples/cfd/lagrangian_mgn/train.py:182]: Maybe add a log_gpu_info function as well to dump the gpu info, namely, number of devices, model, cuda version, etc?

File: examples/cfd/lagrangian_mgn/README.md [hunk 7]
```diff
@@ -22,38 +20,46 @@ steps to maintain physically valid prediction.
...
+system dimensionality of $d = 2$ or $d = 3$, the features are structured
 as follows:
 
-- **Node features**: position (\(d\)), historical velocity (\(t \times d\)),
-                     one-hot encoding of node type (6), wall feature (\(2 \times d\))
-- **Edge features**: displacement (\(d\)), distance (1)
-- **Node target**: acceleration (\(d\))
+- **Node features**:
+  - position ($d$)
+  - historical velocity ($t \times d$)
```
[the remark this answers is not in the document]
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/README.md:58]: Done.

File: examples/cfd/lagrangian_mgn/README.md [hunk 8]
```diff
@@ -65,54 +71,88 @@ a small amount of noise is added during training.
...
+To train the model with the Water dataset, run:
 
 ```bash
-wandb_key: <your_api_key>
+python train.py +experiment=water data.data_dir=/data/Water
 ```
 
-The URL to the dashboard will be displayed in the terminal after the run is launched.
-Alternatively, the logging utility in `train.py` can be switched to MLFlow.
+Progress and loss logs can be monitored using Weights & Biases. To activate that,
```
[the remark this answers is not in the document]
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/README.md:121]: For now we don't have any - but it should be easy to add MLFlow (we already have it in Modulus, just need to properly plug into this code) or TensorBoard. Feel free to take this as a task for yourself.

File: examples/cfd/lagrangian_mgn/conf/config.yaml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+hydra:
+  run:
+    dir: ${output}
+  output_subdir: hydra  # Default is .hydra which causes files not being uploaded in W&B.
+
+# Dimensionality of the problem (2D or 3D).
+dim: 2
+
+# Main output directory.
+output: outputs
```
[the remark this answers is not in the document]
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/conf/config.yaml:31]: With Hydra you can specify it pretty flexibly. For example, `output` can be set by default to:
```
output: outputs/${now:%Y-%m-%d}/${now:%H-%M-%S}
```
This will create a new directory on each based on current time. Users can do the same via command line.
`job.name` does not provide a unique name either - it defaults to the script name (e.g. `train`) so users will still have to override the output via command line/config.
Also, using `./outputs` is kind of default for other Modulus examples (whether it's good or not is a separate discussion)

File: examples/cfd/lagrangian_mgn/conf/config.yaml [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  dataloader:
+    batch_size: ${..batch_size}
+    shuffle: true
+    num_workers: 8
+    pin_memory: true
+    drop_last: true
+
+test:
+  batch_size: 1
+  device: cuda
```
[the remark this answers is not in the document]
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/conf/config.yaml:69]: No, for train the `DistributedManager` is used to set the device properly. If needed, the device can be set using other mechanisms, like `CUDA_VISIBLE_DEVICES`.

File: examples/cfd/lagrangian_mgn/conf/experiment/sand.yaml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    num_steps: 315  # 320 - ${num_history}
+  valid:
+    num_samples: 30
+    num_steps: 100
+  test:
+    num_samples: 30
+    num_steps: 100
+
+model:
+  input_dim_nodes: 25  # 9 node types instead of 6.
```
[the remark this answers is not in the document]
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/conf/experiment/sand.yaml:42]: Yes, `mgn_2d` is used as a base in the `sand` experiment config. But since Sand dataset has 9 node types instead of 6, model input dims need to be overridden.

File: examples/cfd/lagrangian_mgn/inference.py [hunk 12]
```diff
@@ -335,48 +323,51 @@ def plot_error(self, pred, target):
...
 
     # plot
     if cfg.dim == 2:
         rollout.init_animation2d(index=0)
         ani = animation.FuncAnimation(
             rollout.fig,
             rollout.animate2d,
-            frames=(cfg.num_test_time_steps - 5) // cfg.frame_skip,
-            interval=cfg.frame_interval,
+            frames=(cfg.data.test.num_steps - 5) // cfg.inference.frame_skip,
```
[the remark this answers is not in the document]
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/inference.py:358]: `inference.py` will be refactored in the upcoming PR, this PR contains only bare minimum to make it work with new config structure.

File: examples/cfd/lagrangian_mgn/train.py (hunk 3 above)
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/train.py:51]: It's currently not used and I don't think we need it. But I need to confirm with the original author of the code to understand why it was added in the first place.

File: examples/cfd/lagrangian_mgn/train.py (hunk 4 above)
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/train.py:205]: `sum` is faster on lists than `np.sum` (which is faster on numpy arrays but that's not what we have here).

File: examples/cfd/lagrangian_mgn/train.py (hunk 5 above)
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/train.py:203]: This code is refactored in my upcoming PR, please wait.

File: examples/cfd/lagrangian_mgn/train.py (hunk 6 above)
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/train.py:182]: Done.

File: examples/cfd/lagrangian_mgn/README.md (hunk 2 above)
Comment (Alexey-Kamenev) [reply examples/cfd/lagrangian_mgn/README.md:145]: Not at the moment, but it's a great feature! I'm going to add it to the list of tasks.

File: examples/cfd/lagrangian_mgn/README.md [hunk 13]
```diff
@@ -1,11 +1,9 @@
...
-This is an example of Meshgraphnet for particle-based simulation on the
-water dataset based on
-<https://github.com/google-deepmind/deepmind-research/tree/master/learning_to_simulate>
-in PyTorch.
-It demonstrates how to train a Graph Neural Network (GNN) for evaluation
-of the Lagrangian fluid.
+This is an example of MeshGraphNet for particle-based simulation, based on the
+[Learning to Simulate](https://sites.google.com/view/learning-to-simulate/)
+work. It demonstrates how to use Modulus to train a Graph Neural Network (GNN)
+to simulate Lagrangian fluids, solids, and deformable materials.
```
[the remark this answers is not in the document]
Comment (hakhondzadeh) [reply examples/cfd/lagrangian_mgn/README.md:6]: Maybe worthwhile to add another experiment with deformable solid example too as opposed to discrete granular material (sand)!  - I mean in future, not this PR.

**Review by hakhondzadeh — APPROVED** [review]
